include/components/config_parser.hpp

```diff
@@ -0,0 +1,251 @@
+#pragma once
+
+#include <set>
+
+#include "common.hpp"
+#include "errors.hpp"
+#include "components/logger.hpp"
+#include "components/config.hpp"
+#include "utils/file.hpp"
+#include "utils/string.hpp"
+
+POLYBAR_NS
+
+DEFINE_ERROR(parser_error);
+
+/**
+ * \brief Exception object for syntax errors
+ *
+ * Contains filepath and line number where syntax error was found
+ */
+class syntax_error : public parser_error {
+
+  public:
+    /**
+     * Default values are used when the thrower doesn't know the position.
+     * parse_line has to catch, set the proper values and rethrow
+     */
+    explicit syntax_error(const string& msg, const string& file = "", int line_no = -1)
+      : parser_error(file + ":" + to_string(line_no) + ": " + msg), msg(msg) {}
+
+    const string get_msg() { return msg; };
+
+  private:
+    const string& msg;
+};
+
+class invalid_name_error : public syntax_error {
+  public:
+    /**
+     * type is either Header or Key
+     */
+    invalid_name_error(const string& type, const string& name)
+      : syntax_error(type + " '" + name + "' contains forbidden characters.") {}
+};
+
+/**
+ * \brief All different types a line in a config can be
+ */
+enum line_type {KEY, HEADER, COMMENT, EMPTY, UNKNOWN};
```

Comment:
Now we always have to write `line_type::X` :sob: 

But fair enough ^^

Reply:
> Now we always have to write `line_type::X` sob
> 
> But fair enough ^^

I agree, when we'll migrate to C++20, we could use [this](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1099r4.html). :stuck_out_tongue_closed_eyes: 
But that's not until a while since we aren't even using c++17 ^^

BTW what are our conditions to migrate to C++17 (sometimes it could be useful to me)?

Reply:
> Now we always have to write `line_type::X` sob
> 
> But fair enough ^^

I agree, when we'll migrate to C++20, we could use [this](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1099r4.html). :stuck_out_tongue_closed_eyes: 
But that's not until a while since we aren't even using c++17 ^^

BTW what are our conditions to migrate to C++17 (sometimes it could be useful to me)?

Reply:
> BTW what are our conditions to migrate to C++17 (sometimes it could be useful to me)?

Probably once it's no longer experimental in gcc and most popular distros have that new gcc version. We don't want to cut off distros with older compilers.

Reply:
Make sense thanks for your answer. ;)